torchacc/dist/fsdp.py

```diff
@@ -202,3 +204,151 @@ def clip_grad_norm_(self, max_grad_norm):
+                    _, full_params = optim_utils._unflatten_optim_params(
+                        tensor_buffer, layer_name, param_names, param_shapes,
+                        param_numels)
+
+                    for fn, fp in zip(full_names, full_params):
+                        if cpu_offload:
+                            xm.mark_step()
+                            unflat_state_dict[fn][state_name] = fp.cpu()
+                        else:
+                            unflat_state_dict[fn][state_name] = fp
+
+                # all ranks save consolidate optim_state_dict
+                if not rank0_only:
+                    _, full_params = optim_utils._unflatten_optim_params(
+                        tensor_buffer, layer_name, param_names, param_shapes,
+                        param_numels)
+
+                    for fn, fp in zip(full_names, full_params):
+                        if cpu_offload:
+                            xm.mark_step()
+                            unflat_state_dict[fn][state_name] = fp.cpu()
+                        else:
+                            unflat_state_dict[fn][state_name] = fp
+        consolidate_optim_state_dict['state'] = unflat_state_dict
+
+        return consolidate_optim_state_dict
+
+    def load_optim_state_dict(self,
+                              optim_state_dict: Dict[str, Any],
+                              optim: torch.optim.Optimizer,
+                              full_state_dict: str = 'FULL_STATE_DICT',
+                              rank0_only: bool = True) -> Dict[str, Any]:
+        # we only support FULL_STATE_DICT and flatten parameters now
+        assert full_state_dict, 'FULL_STATE_DICT'
+        assert self.model.flatten_parameters, True
+        shard_meta_data = self.model.get_shard_metadata()
+        unflat_optim_state = optim_state_dict
+        flat_optim_state: Dict[str, Any] = {'state': {}, 'param_groups': {}}
+
+        # we only need to flatten and shard the state_dict
+        '''
+        if not rank0_only:
+            for idx, (layer_name, layer_params) in enumerate(shard_meta_data['flatten_info'].items()):
+                param_names, _, _ = layer_params
+                full_names = optim_utils._get_layer_full_names(layer_name, param_names)
+                flat_value: Dict[str, Any] = {}
+                for state_name in unflat_state[full_names[0]].keys():
+                    flat_tensor = optim_utils._flatten_optim_state(unflat_state, state_name, full_names)
+                    if flat_tensor.dim() != 0:
+                        flat_value[state_name] = self.model._get_shard(flat_tensor)
+                
+                flat_optim_state['state'][idx] = flat_value
+                xm.mark_step()
+        '''
+        # broadcast on global ranks instead of sharding_groups
+        unflat_optim_state = optim_utils._broadcast_processed_state(
+            unflat_optim_state, xm.get_ordinal(), xm.xrt_world_size())
```

Comment:
why broadcast on global ranks?

Reply:
> why broadcast on global ranks?

In torchacc's ut, we init a group with full ranks, and then create a new group with new_ranks and init a fsdp model. however, all ranks will go into the function. If we do broadcast with sharing_groups here, only group with rank0 will receive the state_dict information correction, whereas other sharding_groups receive nothing and go wrong.

Reply:
if we have a sharding groups: [[0, 1], [2, 3]] and use rank0_only strategy, only rank 0 have the full information of optim state_dict, rank 1, 2, 3 both load with a empty dict. Then sharding group [2, 3] will receive no information after broadcast and go wrong later.